src/signing/petition_types/petition_entity.rs

```diff
@@ -0,0 +1,557 @@
+    {
+        let t = self.on_list(FactorListKind::Threshold, &r#do);
+        let o = self.on_list(FactorListKind::Override, &r#do);
+        combine(t, o)
+    }
+
+    fn both_void<F, R>(&self, r#do: F)
+    where
+        F: Fn(&PetitionFactors) -> R,
+    {
+        self.both(r#do, |_, _| ())
+    }
+
+    pub fn skipped_factor_source_if_relevant(&self, factor_source_id: &FactorSourceIDFromHash) {
+        self.both_void(|l| l.skip_if_references(factor_source_id, true));
+    }
+
+    /// # Panics
+    /// Panics if this factor source has already been skipped or signed with.
+    ///
+    /// Or panics if the factor source is not known to this petition.
+    pub fn add_signature(&self, signature: HDSignature) {
+        self.both(|l| l.add_signature_if_relevant(&signature), |t, o| {
+            match (t, o) {
+                (Some(true), Some(true)) => {
+                    unreachable!("Matrix of FactorInstances does not allow for a factor to be present in both threshold and override list, thus this will never happen.")
+                }
+                (Some(false), Some(false)) => panic!("Factor source not found in any of the lists."),
+                (None, None) => panic!("Programmer error! Must have at least one factors list."), 
+                _ => (),
+            }
+        })
+    }
+
+    pub fn invalid_transactions_if_skipped_factors(
+        &self,
+        factor_source_ids: IndexSet<FactorSourceIDFromHash>,
+    ) -> IndexSet<InvalidTransactionIfSkipped> {
+        let skip_status = self.status_if_skipped_factors(factor_source_ids);
+        match skip_status {
+            PetitionFactorsStatus::Finished(finished_reason) => match finished_reason {
+                PetitionFactorsStatusFinished::Fail => {
+                    let intent_hash = self.intent_hash.clone();
+                    let invalid_transaction =
+                        InvalidTransactionIfSkipped::new(intent_hash, vec![self.entity.clone()]);
+                    IndexSet::from_iter([invalid_transaction])
+                }
+                PetitionFactorsStatusFinished::Success => IndexSet::new(),
+            },
+            PetitionFactorsStatus::InProgress => IndexSet::new(),
+        }
+    }
+
+    pub fn status_if_skipped_factors(
+        &self,
+        factor_source_ids: IndexSet<FactorSourceIDFromHash>,
+    ) -> PetitionFactorsStatus {
+        let simulation = self.clone();
+        for factor_source_id in factor_source_ids.iter() {
+            simulation
+                .did_skip_if_relevant(factor_source_id, true)
+                .unwrap();
+        }
+        simulation.status()
+    }
+
+    pub fn did_skip_if_relevant(
+        &self,
+        factor_source_id: &FactorSourceIDFromHash,
+        simulated: bool,
+    ) -> Result<()> {
+        self.both_void(|p| p.did_skip_if_relevant(factor_source_id, simulated));
+        Ok(())
+    }
+
+    pub fn status(&self) -> PetitionFactorsStatus {
```

Comment:
Any reason this whole function does not use `both`?

Reply:
Nope! well spotted! I just forgot to use it, fixed in https://github.com/radixdlt/sargon-mfa/pull/6/commits/7426b330f058e03094a7e32328c0cc53299d12a7